bookserver/crud.py

```diff
@@ -118,10 +127,118 @@ async def fetch_course(course_name: str) -> CoursesValidator:
     return CoursesValidator.from_orm(res.scalars().first())
 
 
+async def create_course(course_info: CoursesValidator) -> CoursesValidator:
+    new_course = Courses(**course_info.dict())
+    async with async_session() as session:
+        res = session.add(new_course)
+        await session.commit()
+
+    return CoursesValidator.from_orm(res) if res else None
+
+
+# auth_user
+# ---------
 async def fetch_user(user_name: str) -> AuthUserValidator:
     query = select(AuthUser).where(AuthUser.username == user_name)
     async with async_session() as session:
         res = await session.execute(query)
         rslogger.debug(f"res = {res}")
     user = res.scalars().first()
     return AuthUserValidator.from_orm(user) if user else None
+
+
+async def create_user(user: AuthUserValidator) -> int:
+    """
+    The given user will have the password in plain text.  First we will hash
+    the password then add this user to the database.
+    """
+    new_user = AuthUser(**user.dict())
+    print(settings.web2py_private_key)
+    crypt = CRYPT(key=settings.web2py_private_key, salt=True)
+    new_user.password = str(crypt(user.password)[0])
+    res = None
+    try:
+        async with async_session() as session:
+            res = session.add(new_user)
+            await session.commit()
+    except IntegrityError:
+        rslogger.error("Failed to add a duplicate user")
+
+    return AuthUserValidator.from_orm(res) if res else None
+
+
+# instructor_courses
+# ------------------
+async def fetch_instructor_courses(
+    instructor_id: int, course_id: Optional[int] = None
+) -> List[CourseInstructorValidator]:
+    """
+    return a list of courses for which the given userid is an instructor.
+    If the optional course_id value is included then retur the row for that
+    course to verify that instructor_id is an instructor for course_id
+    """
+
+    query = select(CourseInstructor)
+    if course_id is not None:
+        query = query.where(
+            and_(
+                CourseInstructor.instructor == instructor_id,
+                CourseInstructor.course == course_id,
+            )
+        )
+    else:
+        query = query.where(CourseInstructor.instructor == instructor_id)
+    async with async_session() as session:
+        res = await session.execute(query)
+
+    course_list = [
+        CourseInstructorValidator.from_orm(x) for x in res.scalars().fetchall()
+    ]
+    return course_list
+
+
+async def create_initial_courses_users():
+    if settings.drop_tables == "Yes":
```

Comment:
This is a bit more work, but I think it would be neat if courses was populated based on the subdirectories found in `setting.book_path` (unless that course entry already existed). Then, users could download a book to the appropriate location, restart the server, and it's ready to go!

Reply:
Interesting idea... I've been leaning the other way and wanting to make the book stuff more database driven and less reliant on the names of directories and the presence of conf.py files in things that are not really packages.

Reply:
What I'm hoping for is a simple end-user experience: unzip a book, put it in the right place, and it just works.  I'm concerned the database is less accessible than this process.

Reply:
I agree that we want to make it easy.  But the number of people that run their own servers is pretty small!  And... there is a new funded path for making custom books easy!  The github webhooks project.  I have at least partial funding for that through UT Austin.  So there will be a webpage that allows you to select a github repo and use that book for your course.

Can we agree that this should not hold up merging the PR?

Reply:
Yes, I agree.